File: Kryptos/Model/SignalWire.swift
import Foundation
import LibSignalClient
import CipherCore

enum SignalWire {
    private static let ctx = NullContext()

    static func pairKey(_ a: String, _ b: String) -> Data {
        Data((a <= b ? a + b : b + a).utf8)
    }

    struct Sealed: Sendable {
        let ciphertext: Data
        let type: UInt8
        let deflate: Bool
        let pad: Bool
        let pairKey: Data
    }

    struct Cover: Sendable {
        let text: String
        let hidden: Bool
    }

    static func seal(_ text: String, toFingerprint fp: String, myFingerprint: String,
                     store: PersistentSignalStore, pad: Bool) throws -> Sealed {
        let addr = try ProtocolAddress(name: fp, deviceId: 1)
        let myAddr = try ProtocolAddress(name: myFingerprint, deviceId: 1)
        let raw = Data(text.utf8)
        let compressed = Deflate.compress(raw)
        let deflate = compressed != nil
        let ct = try signalEncrypt(message: Array(deflate ? compressed! : raw), for: addr, localAddress: myAddr,
                                   sessionStore: store, identityStore: store, context: ctx)
        return Sealed(ciphertext: ct.serialize(), type: ct.messageType.rawValue, deflate: deflate,
                      pad: pad, pairKey: pairKey(myFingerprint, fp))
    }

    static func cover(_ sealed: Sealed, stego: StegoLanguage?, mode: StegoMode) throws -> Cover {
        if let language = stego {
            let padded = sealed.pad && StegoWire.fits(ciphertext: sealed.ciphertext.count, padded: true)
            let payload = StegoWire.frame(sealed.ciphertext, type: sealed.type,
                                          deflate: sealed.deflate, padded: padded)
            if payload.count <= TextStego.maxPayloadBytes {
                let cover: String?
                switch mode {
                case .words: cover = TextStego.encode(payload, language: language)
                case .smart: cover = SmartTextStego.encode(payload, language: language)
                case .letters: cover = LetterStego.encode(payload, language: language)
                }
                if let cover { return Cover(text: cover, hidden: true) }
            }
        }
        let token = try WireFormat.wrap(sealed.ciphertext, type: sealed.type, deflate: sealed.deflate,
                                        padded: sealed.pad, pairKey: sealed.pairKey)
        return Cover(text: token, hidden: false)
    }

    static func encrypt(_ text: String, toFingerprint fp: String, myFingerprint: String,
                        store: PersistentSignalStore, stego: StegoLanguage? = nil, mode: StegoMode = .words,
                        pad: Bool = false) throws -> String {
        let sealed = try seal(text, toFingerprint: fp, myFingerprint: myFingerprint, store: store, pad: pad)
        return try cover(sealed, stego: stego, mode: mode).text
    }

    static func decrypt(_ armored: String, fromFingerprint fp: String, myFingerprint: String,
                        store: PersistentSignalStore, stego precomputed: Data?? = nil) throws -> String {
        let addr = try ProtocolAddress(name: fp, deviceId: 1)
        let myAddr = try ProtocolAddress(name: myFingerprint, deviceId: 1)
        let hidden = { precomputed ?? stegoPayload(armored) }

        if let (type, deflate, body) = WireFormat.unwrap(armored, pairKey: pairKey(myFingerprint, fp)) {
            do {
                let plain = try signalDecryptBytes(type: type, body: body, addr: addr, myAddr: myAddr, store: store)
                return try inflate(plain, deflate: deflate)
            } catch {
                guard let payload = hidden() else { throw error }
                return try decryptStego(payload, addr: addr, myAddr: myAddr, store: store)
            }
        }

        guard let payload = hidden() else { throw CipherError.notAKryptosMessage }
        return try decryptStego(payload, addr: addr, myAddr: myAddr, store: store)
    }

    static let maxStegoInputChars = 1_000_000

    static func stegoPayload(_ armored: String) -> Data? {
        guard armored.utf16.count <= maxStegoInputChars else { return nil }
        return TextStego.decode(armored) ?? SmartTextStego.decode(armored) ?? LetterStego.decode(armored)
    }

    private static func inflate(_ plain: Data, deflate: Bool) throws -> String {
        guard deflate else { return String(decoding: plain, as: UTF8.self) }
        guard let data = Deflate.decompress(plain) else { throw CipherError.decryptionFailed }
        return String(decoding: data, as: UTF8.self)
    }

    private static func decryptStego(_ payload: Data, addr: ProtocolAddress, myAddr: ProtocolAddress,
                                     store: PersistentSignalStore) throws -> String {
        guard let framed = StegoWire.unframe(payload) else {
            throw StegoWire.carriesUnknownFlags(payload) ? CipherError.unsupportedFormat : CipherError.notAKryptosMessage
        }
        let plain = try signalDecryptBytes(type: framed.type, body: framed.body,
                                           addr: addr, myAddr: myAddr, store: store)
        return try inflate(plain, deflate: framed.deflate)
    }

    private static func signalDecryptBytes(type: UInt8, body: Data, addr: ProtocolAddress, myAddr: ProtocolAddress,
                                           store: PersistentSignalStore) throws -> Data {
        if type == CiphertextMessage.MessageType.preKey.rawValue {
            return try signalDecryptPreKey(message: PreKeySignalMessage(bytes: body),
                                           from: addr, localAddress: myAddr,
                                           sessionStore: store, identityStore: store,
                                           preKeyStore: store, signedPreKeyStore: store, kyberPreKeyStore: store, context: ctx)
        }
        return try signalDecrypt(message: SignalMessage(bytes: body), from: addr, to: myAddr,
                                 sessionStore: store, identityStore: store, context: ctx)
    }

}

File: android/app/src/main/java/com/kryptos/android/signal/SignalWire.kt
package com.kryptos.android.signal

import com.kryptos.android.core.CipherException
import com.kryptos.android.core.Deflate
import com.kryptos.android.core.LetterStego
import com.kryptos.android.core.SmartTextStego
import com.kryptos.android.core.StegoMode
import com.kryptos.android.core.StegoWire
import com.kryptos.android.core.StegoLanguage
import com.kryptos.android.core.TextStego
import com.kryptos.android.core.WireFormat
import org.signal.libsignal.protocol.SessionCipher
import org.signal.libsignal.protocol.SignalProtocolAddress
import org.signal.libsignal.protocol.message.CiphertextMessage
import org.signal.libsignal.protocol.message.PreKeySignalMessage
import org.signal.libsignal.protocol.message.SignalMessage
import org.signal.libsignal.protocol.state.SignalProtocolStore

object SignalWire {

    fun pairKey(a: String, b: String): ByteArray =
        (if (a <= b) a + b else b + a).toByteArray(Charsets.UTF_8)

    fun encrypt(
        text: String,
        toFingerprint: String,
        myFingerprint: String,
        store: SignalProtocolStore,
        stego: StegoLanguage? = null,
        mode: StegoMode = StegoMode.WORDS,
        pad: Boolean = false,
    ): String {
        val addr = SignalProtocolAddress(toFingerprint, 1)
        val myAddr = SignalProtocolAddress(myFingerprint, 1)
        val cipher = SessionCipher(store, myAddr, addr)

        if (stego != null) {
            val raw = text.toByteArray(Charsets.UTF_8)
            val compressed = Deflate.compress(raw)
            val deflate = compressed != null
            val ct = cipher.encrypt(if (deflate) compressed!! else raw)
            val serialized = ct.serialize()
            val padded = pad && StegoWire.fits(serialized.size, true)
            val payload = StegoWire.frame(serialized, ct.type, deflate, padded)
            if (payload.size <= TextStego.MAX_PAYLOAD_BYTES) {
                return when (mode) {
                    StegoMode.WORDS -> TextStego.encode(payload, stego)
                    StegoMode.SMART -> SmartTextStego.encode(payload, stego)
                    StegoMode.LETTERS -> LetterStego.encode(payload, stego)
                }
            }
            return WireFormat.wrap(serialized, ct.type, deflate, pad, pairKey(myFingerprint, toFingerprint))
        }

        val plaintext = text.toByteArray(Charsets.UTF_8)
        val compressed = Deflate.compress(plaintext)
        val deflate = compressed != null
        val ct = cipher.encrypt(if (deflate) compressed!! else plaintext)
        return WireFormat.wrap(ct.serialize(), ct.type, deflate, pad, pairKey(myFingerprint, toFingerprint))
    }

    fun decrypt(armored: String, fromFingerprint: String, myFingerprint: String, store: SignalProtocolStore): String {
        val addr = SignalProtocolAddress(fromFingerprint, 1)
        val myAddr = SignalProtocolAddress(myFingerprint, 1)
        val cipher = SessionCipher(store, myAddr, addr)

        val unwrapped = WireFormat.unwrap(armored, pairKey(myFingerprint, fromFingerprint))
        if (unwrapped != null) {
            try {
                val raw = signalDecrypt(cipher, unwrapped.type, unwrapped.body)
                val data = if (unwrapped.deflate) Deflate.decompress(raw) ?: ByteArray(0) else raw
                return String(data, Charsets.UTF_8)
            } catch (e: Exception) {
                val fallback = stegoPayload(armored) ?: throw e
                return decryptStego(cipher, fallback)
            }
        }

        val payload = stegoPayload(armored) ?: throw CipherException(CipherException.Kind.NOT_A_KRYPTOS_MESSAGE)
        return decryptStego(cipher, payload)
    }

    private const val MAX_STEGO_INPUT_CHARS = 1_000_000

    private fun stegoPayload(armored: String): ByteArray? {
        if (armored.length > MAX_STEGO_INPUT_CHARS) return null
        return TextStego.decode(armored) ?: SmartTextStego.decode(armored) ?: LetterStego.decode(armored)
    }

    private fun decryptStego(cipher: SessionCipher, payload: ByteArray): String {
        val framed = StegoWire.unframe(payload) ?: throw CipherException(
            if (StegoWire.carriesUnknownFlags(payload)) CipherException.Kind.UNSUPPORTED_FORMAT
            else CipherException.Kind.NOT_A_KRYPTOS_MESSAGE
        )
        val raw = signalDecrypt(cipher, framed.type, framed.body)
        val data = if (framed.deflate) Deflate.decompress(raw) ?: ByteArray(0) else raw
        return String(data, Charsets.UTF_8)
    }

    private fun signalDecrypt(cipher: SessionCipher, type: Int, body: ByteArray): ByteArray =
        if (type == CiphertextMessage.PREKEY_TYPE) {
            cipher.decrypt(PreKeySignalMessage(body))
        } else {
            cipher.decrypt(SignalMessage(body))
        }
}
